### ISSUE TITLE
[Core feature] Structured dataset in flytekit/flyte

### ISSUE THREAD

**Issue by pingsutw** [flyteorg/flyte#2074]
### Motivation: Why do you think this is important?

# TL;DR
Introduces a new type and user experience for working with tabular data in Flyte. Please see the [design doc](https://docs.google.com/document/d/1-t8C4vW_RhrJoC5Ju16fmqpyOvsCRb3XmulLLo5VX0E/) for the complete background.

* The new type is the `StructuredDataset`. (see the [IDL PR](https://github.com/flyteorg/flyteidl/pull/227)) We thought the name sufficiently generic and differentiated from `FlyteSchema`, which we're also keeping around for the time being (see more in the Migration section below). The main difference between the `FlyteSchema` and `StructuredDataset` types themselves is that columns in the latter are just named `LiteralType`s. That is, columns are no longer constrained to be purely primitives.

**Note:**
The `[kwtype(...)]` syntax used for specifying columns and storage formats will be changing before the next non-beta release of flytekit, so if you try this out now, keep in mind you'll have some minor updates to make in a couple weeks.

**Note:**
To handle the new type, you will need to be on Propeller version [v0.16.14](https://github.com/flyteorg/flytepropeller/releases/tag/v0.16.14) and Admin version [v0.6.78](https://github.com/flyteorg/flyteadmin/releases/tag/v0.6.78) or later.

### Goal: What should the final outcome look like, ideally?

* The contributor-experience has also changed substantially from `FlyteSchema` to something we hope is more straightforward and easier to understand, but we've tried to keep the user-experience as similar as possible.

### Describe alternatives you've considered 

None

###  Propose: Link/Inline OR Additional context 

- https://github.com/flyteorg/flytekit/pull/785
- https://github.com/flyteorg/flytekit/pull/789
- https://github.com/flyteorg/flyteidl/pull/227
- https://github.com/flyteorg/flytepropeller/pull/384
- https://github.com/flyteorg/flyteplugins/pull/233
- https://github.com/flyteorg/flytesnacks/pull/521
- https://github.com/flyteorg/flytekit/pull/822

## TODO
- [ ] Support Athena plugin to output structured dataset
- [ ] Support Snowflake plugin to output structured dataset
- [x] Figure out the backwards compatibility story
- [x] Change StructuredDataset[my_cols] to Annotated[StructuredDataset, my_cols]
- [x] Column subsetting
- [ ] Add new fsspec[aws] and fsspec[gcp] flytekit plugins
- [x] Docs and slides
- [x] Flytesnack example

### Are you sure this issue hasn't been raised already?
- [X] Yes

Relates to: https://github.com/flyteorg/flyte/issues/523 though

### Have you read the Code of Conduct?

- [X] Yes

**Comment by EngHabu** [comment]
Done
